README.md
# os-monitor

[![NPM](https://nodei.co/npm/os-monitor.png)](https://nodei.co/npm/os-monitor/)

A very simple monitor for the built-in `os` module in Node.js.

Allows you to observe some OS parameters, such as free memory available or load average.


## Installation

    npm install os-monitor


# Synopsis

```javascript
const monitor = require("os-monitor");


// basic usage
monitor.start();

// more advanced usage with configs.
monitor.start({ delay: 3000 // interval in ms between monitor cycles
              , freemem: 1000000000 // freemem under which event 'freemem' is triggered
              , uptime: 1000000 // number of secs over which event 'uptime' is triggered
              , critical1: 0.7 // loadavg1 over which event 'loadavg1' is triggered
              , critical5: 0.7 // loadavg5 over which event 'loadavg5' is triggered
              , critical15: 0.7 // loadavg15 over which event 'loadavg15' is triggered
              , silent: false // set true to mute event 'monitor'
              , stream: false // set true to enable the monitor as a Readable Stream
              , immediate: false // set true to execute a monitor cycle at start()
              });


// define handler that will always fire every cycle
monitor.on('monitor', (event) => {
  console.log(event.type, 'This event always happens on each monitor cycle!');
});

// define handler for a too high 1-minute load average
monitor.on('loadavg1', (event) => {
  console.log(event.type, 'Load average is exceptionally high!');
});

// define handler for a too low free memory
monitor.on('freemem', (event) => {
  console.log(event.type, 'Free memory is very low!');
});

// define a throttled handler, using Underscore.js's throttle function (http://underscorejs.org/#throttle)
monitor.throttle('loadavg5', (event) => {

  // whatever is done here will not happen
  // more than once every 5 minutes(300000 ms)

}, monitor.minutes(5));


// change config while monitor is running
monitor.config({
  freemem: 0.3 // alarm when 30% or less free memory available
});


// stop monitor
monitor.stop();


// check whether monitor is running or not
monitor.isRunning(); // -> true / false


// use as readable stream
monitor.start({stream: true}).pipe(process.stdout);

```

# config options

###  delay

Delay in milliseconds between each monitor cycle. Default: 3000

###  freemem

Amount of memory in bytes under which event 'freemem' is triggered. Can also be a percentage of total memory. Default: 0

###  uptime

Number of seconds over which event 'uptime' is triggered. Default: undefined

###  critical1

Value of 1 minute load average over which event 'loadavg1' is triggered. Default: os.cpus().length

(The load average is a measure of system activity, calculated by the operating system and expressed as a fractional number. As a rule of thumb, the load average should ideally be less than the number of logical CPUs in the system. ref.: [http://nodejs.org/api/os.html#os_os_loadavg](http://nodejs.org/api/os.html#os_os_loadavg "load average"))

###  critical5

Value of 5 minutes load average over which event 'loadavg5' is triggered. Default: os.cpus().length

###  critical15

Value of 15 minutes load average over which event 'loadavg15' is triggered. Default: os.cpus().length

###  silent

Set true to mute event 'monitor'. Default: false

###  stream

Set true to enable the monitor as a [Readable Stream](http://nodejs.org/api/stream.html#stream_class_stream_readable "Readable Stream"). Default: false

###  immediate

Set true to execute a monitor cycle at start(). Default: false


# API

### .version

The `monitor.version` property contains the `os-monitor` version string.

### .start( [options] )

Starts the monitor. Accepts an optional options object.

### .stop( )

Stops the monitor.

### .isRunning( )

Checks whether the monitor is running or not; returns a boolean.

### .config( [options] )

Accepts an optional options object and updates monitor config. Always returns monitor config options.

### .reset( )

Resets monitor config to its default values.

### .on( eventType, handler ), .addListener( eventType, handler )

Adds a listener for the specified event type. Supported events are: 'monitor', 'uptime', 'freemem', 'loadavg1', 'loadavg5', 'loadavg15', 'start', 'stop', 'config', 'reset', 'destroy'.

### .once( eventType, handler )

Adds a one-time listener for the specified event type. This listener is invoked only the next time the event is fired, after which it is removed.

### .throttle( eventType, handler, delay )

Adds a throttled listener, using [Underscore.js's throttle](http://underscorejs.org/#throttle) function. The throttled listener will not be executed more than once every delay milliseconds.

### .unthrottle( eventType, handler )

Removes a throttled listener previously added using `.throttle()`. `handler` must be the original function.

### .when( eventType )

Returns a Promise(or a basic thenable if Promise is not supported) that resolves with an event object when `eventType` is triggered.

### .destroy( )

Permanently stops and disables the monitor.

### .seconds(n), .minutes(n), .hours(n), .days(n)

Convenience methods to get the right amount of milliseconds.
```javascript
monitor.seconds(10); // -> 10000 ms

monitor.minutes(5); // -> 300000 ms

monitor.hours(1); // -> 3600000 ms

monitor.days(1); // -> 86400000 ms

// start with a delay of 5000 ms
monitor.start({ delay: monitor.seconds(5) });

```


## Event object

There is some useful information in the provided event object:

```
{
  type: 'monitor', // event type
  loadavg: [ 0.4599609375, 0.53076171875, 0.4990234375 ], // load average values for 1, 5, 15 minutes
  uptime: 1614056, // os uptime in seconds
  freemem: 241262592, // free memory available in bytes
  totalmem: 2147483648, // total memory available in bytes
  timestamp: 1394766898 // UNIX Timestamp
}
```
All supported events are: 'monitor', 'uptime', 'freemem', 'loadavg1', 'loadavg5', 'loadavg15', 'start', 'stop', 'config', 'reset', 'destroy'.
<em>Note that `os-monitor` is an instance of `EventEmitter`</em>.

Events API docs: [nodejs.org/api/events](http://nodejs.org/api/events.html "Events")


## Using the monitor as a Readable Stream

`os-monitor` can also be used as a [Readable Stream](http://nodejs.org/api/stream.html#stream_class_stream_readable "Readable Stream").

```
monitor.start({ stream: true });


// write to STDOUT
monitor.pipe(process.stdout);


// write to a file
let fs = require('fs'),
    logFile = fs.createWriteStream('/tmp/log.txt', {flags: 'a'});

monitor.pipe(logFile);
```


## Promise / thenable

`os-monitor` supports Promise, async/await: using `.when(eventType)` returns a Promise(or a basic thenable if Promise is not supported).

```
monitor.when('freemem').then(event => {
    // ...
});

async function callback() {
    let event = await monitor.when('uptime');
    // ...
}
```


## Monitor class

Need concurrent monitor instances? The monitor class is available from the os-monitor object:

```
const monitor = require('os-monitor');

let monitor1 = new monitor.Monitor();
let monitor2 = new monitor.Monitor();
let monitor3 = new monitor.Monitor();
```


## Node.js os module

The node `os` built-in module is also available from the os-monitor object:

```
const monitor = require('os-monitor');

let type = monitor.os.type();
let cpus = monitor.os.cpus();
```

Documentation for the `os` module is available [here](http://nodejs.org/api/os.html).


CHANGELOG.md
## v1.2.0
* [[`36e2ed1bc4`](https://github.com/lfortin/node-os-monitor/commit/36e2ed1bc44fa6faacbed63f8e7a78269c82c19f)] - documented .version property (Laurent Fortin)
* [[`6834f8226f`](https://github.com/lfortin/node-os-monitor/commit/6834f8226f1d55cc85ab59ef9bc88ff416e42a7a)] - introduce .when() method (Laurent Fortin)
## v1.1.0
* [[`ca38d7a2de`](https://github.com/lfortin/node-os-monitor/commit/ca38d7a2dea12ee33d559f25099feb115b390fa6)] - introduce .unthrottle() method (Laurent Fortin)


test/spec-runner.js

const assert = require('assert'),
          os = require('os'),
        mock = require('mock-os'),
           _ = require('underscore');

const monitor = require('../os-monitor');
let tester;

beforeEach(() => {
  mock({
    cpus: [{}, {}],
    loadavg: [2, 2, 2],
    freemem: 1000000,
    totalmem: 100000000,
    uptime: 1000000,
  });
  tester = new monitor.Monitor();
});
afterEach(() => {
  tester.destroy();
  mock.restore();
});

describe('Monitor class', function() {
  it('should have default instance', async () => {
    assert.ok(monitor instanceof monitor.Monitor);
  });
  it('should create new instances', async () => {
    assert.ok(tester instanceof monitor.Monitor);
  });
});
describe('API signature', function() {
  it('should have valid API signature', async () => {
    // API signature
    assert.ok(tester._read, "internal ._read() method expected");
    assert.ok(tester.version, "version property expected");
    assert.ok(tester.sendEvent, ".sendEvent() method expected");
    assert.ok(tester.start, ".start() method expected");
    assert.ok(tester.stop, ".stop() method expected");
    assert.ok(tester.reset, ".reset() method expected");
    assert.ok(tester.destroy, ".destroy() method expected");
    assert.ok(tester.config, ".config() method expected");
    assert.ok(tester.isRunning, ".isRunning() method expected");
    assert.ok(tester.throttle, ".throttle() method expected");
    assert.ok(tester.unthrottle, ".unthrottle() method expected");
    assert.ok(tester.when, ".when() method expected");
    assert.ok(tester.seconds, ".seconds() method expected");
    assert.ok(tester.minutes, ".minutes() method expected");
    assert.ok(tester.hours, ".hours() method expected");
    assert.ok(tester.days, ".days() method expected");
    assert.ok(tester.Monitor, "Monitor class expected");
    assert.ok(tester.Thenable, "Thenable class expected");
    assert.ok(tester.os, "os object reference expected");
    assert.ok(tester._, "_ object reference expected");
    assert.ok(tester.constants, "constants object expected");
    //assert.ok(monitor.whetever, "whatever expected");
  });
  it('should have readonly version', async () => {
    let version = tester.version;
    tester.version = '0.0.0';
    assert.strictEqual(tester.version, version);
  });
});
describe('event emitter', function() {
  it('should emit start event', (done) => {
    tester.on('start', event => {
      assert.strictEqual(event.type, tester.constants.events.START);
      done();
    });
    tester.start();
  });
  it('should emit stop event', (done) => {
    tester.start({
      immediate: true,
    });
    tester.on('stop', event => {
      assert.strictEqual(event.type, tester.constants.events.STOP);
      done();
    });
    tester.stop();
  });
  it('should emit config event', (done) => {
    tester.on('config', event => {
      assert.strictEqual(event.type, tester.constants.events.CONFIG);
      assert.strictEqual(event.options.stream, true);
      done();
    });
    tester.config({stream: true});
  });
  it('should emit reset event', (done) => {
    tester.on('reset', event => {
      assert.strictEqual(event.type, tester.constants.events.RESET);
      done();
    });
    tester.reset();
  });
  it('should emit destroy event', (done) => {
    tester.on('destroy', event => {
      assert.strictEqual(event.type, tester.constants.events.DESTROY);
      done();
    });
    tester.destroy();
  });
  it('should emit monitor event', (done) => {
    tester.on('monitor', event => {
      assert.strictEqual(event.type, tester.constants.events.MONITOR);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
    tester.start({
      immediate: true,
    });
  });
  it('should not emit monitor event', async () => {
    tester.on('monitor', event => {
      assert.fail('should not emit monitor event');
    });
    tester.start({
      immediate: true,
      silent: true,
    });
  });
  it('should emit loadavg1 event', (done) => {
    tester.start({
      critical1: 1,
      immediate: true,
    });
    tester.on('loadavg1', event => {
      assert.strictEqual(event.type, tester.constants.events.LOADAVG1);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
  });
  it('should emit loadavg5 event', (done) => {
    tester.start({
      critical5: 1,
      immediate: true,
    });
    tester.on('loadavg5', event => {
      assert.strictEqual(event.type, tester.constants.events.LOADAVG5);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
  });
  it('should emit loadavg15 event', (done) => {
    tester.start({
      critical15: 1,
      immediate: true,
    });
    tester.on('loadavg15', event => {
      assert.strictEqual(event.type, tester.constants.events.LOADAVG15);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
  });
  it('should emit freemem event', (done) => {
    tester.start({
      freemem: 10000000,
      immediate: true,
    });
    tester.on('freemem', event => {
      assert.strictEqual(event.type, tester.constants.events.FREEMEM);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
  });
  it('should emit uptime event', (done) => {
    tester.start({
      uptime: 10000,
      immediate: true,
    });
    tester.on('uptime', event => {
      assert.strictEqual(event.type, tester.constants.events.UPTIME);
      assert.ok(event.loadavg);
      assert.ok(event.freemem);
      assert.ok(event.totalmem);
      assert.ok(event.uptime);
      assert.ok(event.timestamp);
      done();
    });
  });
});
describe('readable stream', function() {
  it('should read data', async () => {
    tester.pause();
    tester.start({
      stream: true,
      immediate: true,
    });
    assert.ok(tester.read());
  });
  it('should stop buffering if full', async () => {
    tester.pause();
    tester.push(Buffer.alloc(200000));
    tester.start({
      stream: true,
      immediate: true,
    });
    assert.strictEqual(tester.push('hello'), false);
  });
});
describe('cycles', function() {
  it('should execute multiple cycles', (done) => {
    let cycles = 0;
    tester.start({delay: 25});
    tester.on('monitor', () => {
      cycles++;
    });
    tester.on('stop', () => {
      assert.ok(cycles > 2);
      done();
    });
    setTimeout(() => tester.stop(), 100);
  });
});
describe('.throttle()', function() {
  it('should register throttled handler', (done) => {
    let cycles = 0;
    tester.throttle('monitor', event => {
      cycles++;
    }, 50);
    tester.on('stop', event => {
      assert.ok(cycles < 3);
      done();
    });
    tester.start({delay: 10});
    setTimeout(() => {
      tester.stop();
    }, 100);
  });
});
describe('.unthrottle()', function() {
  it('should remove throttled handler', (done) => {
    let cycles = 0;
    let handler = (event) => {
      cycles++;
    }
    tester.throttle('monitor', handler, 50);
    tester.unthrottle('monitor', handler);
    tester.on('stop', event => {
      assert.strictEqual(cycles, 0);
      done();
    });
    tester.start({delay: 10});
    setTimeout(() => {
      tester.stop();
    }, 100);
  });
});
describe('.when()', function() {
  it('should return a promise that resolves', async () => {
    tester.start({delay: 10});
    let event = await tester.when('monitor');
    assert.strictEqual(event.type, 'monitor');
    assert.ok(tester.when('monitor') instanceof Promise);
  });
  it('should return a thenable that resolves', async () => {
    let resolve = Promise.resolve;
    delete Promise.resolve;
    tester.start({delay: 10});
    let event = await tester.when('monitor');
    assert.strictEqual(event.type, 'monitor');
    assert.ok(tester.when('monitor') instanceof tester.Thenable);
    Promise.resolve = resolve;
  });
});
describe('Thenable class', function() {
  it('should create a thenable that resolves', (done) => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.then(result => {
      assert.strictEqual(result, value);
      done();
    });
    deferred.resolve(value);
  });
  it('should create a thenable that rejects', (done) => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.catch(reason => {
      assert.strictEqual(reason, value);
      done();
    });
    deferred.reject(value);
  });
  it('should handle multiple resolutions', (done) => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.then(result => {
      assert.strictEqual(result, value);
      done();
    });
    deferred.resolve(value);
    deferred.resolve();
    deferred.reject(12345);
  });
  it('should handle multiple rejections', (done) => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.catch(reason => {
      assert.strictEqual(reason, value);
      done();
    });
    deferred.reject(value);
    deferred.reject();
    deferred.resolve(12345);
  });
  it('should handle early resolution', async () => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.resolve(value);
    deferred.resolve(123);
    deferred.reject(12345);
    let result = await deferred;
    assert.strictEqual(result, value);
  });
  it('should handle early rejection', async () => {
    let deferred = new tester.Thenable();
    let value = {};
    deferred.reject(value);
    deferred.resolve(123);
    deferred.reject(12345);
    assert.rejects(async () => {
      await deferred;
    });
  });
});
describe('.destroy()', function() {
  it('should permanently stop', async () => {
    tester.destroy();
    assert.throws(() => {
      tester.start();
    });
  });
});
describe('convenience methods', function() {
  describe('.seconds(), .minutes(), .hours(), .days()', function() {
    it('should convert seconds into ms', async () => {
      assert.strictEqual(tester.seconds(1), 1000);
      assert.strictEqual(tester.seconds(5), 5000);
    });
    it('should convert minutes into ms', async () => {
      assert.strictEqual(tester.minutes(1), 60000);
      assert.strictEqual(tester.minutes(5), 300000);
    });
    it('should convert hours into ms', async () => {
      assert.strictEqual(tester.hours(1), 3600000);
      assert.strictEqual(tester.hours(10), 36000000);
    });
    it('should convert days into ms', async () => {
      assert.strictEqual(tester.days(1), 86400000);
      assert.strictEqual(tester.days(2), 172800000);
    });
    it('should fail if not a number', async () => {
      assert.throws(() => {
        tester._sanitizeNumber('not a number');
      });
      assert.throws(() => {
        tester.seconds('not a number');
      });
      assert.throws(() => {
        tester.minutes('not a number');
      });
      assert.throws(() => {
        tester.hours('not a number');
      });
      assert.throws(() => {
        tester.days('not a number');
      });
    });
    it('should fail if < 0', async () => {
      assert.throws(() => {
        tester._sanitizeNumber(-1);
      });
      assert.throws(() => {
        tester.seconds(-1);
      });
      assert.throws(() => {
        tester.minutes(-1);
      });
      assert.throws(() => {
        tester.hours(-1);
      });
      assert.throws(() => {
        tester.days(-1);
      });
    });
    it('should fail if > 2147483648', async () => {
      assert.throws(() => {
        tester._sanitizeNumber(2147483649);
      });
      assert.throws(() => {
        tester.seconds(2147483649);
      });
      assert.throws(() => {
        tester.minutes(2147483649);
      });
      assert.throws(() => {
        tester.hours(2147483649);
      });
      assert.throws(() => {
        tester.days(2147483649);
      });
    });
  });
});
describe('chainable methods', function() {
  it('should support chainable methods', async () => {
    tester.start()
          .reset()
          .stop()
          .throttle('start', _.noop)
          .unthrottle('start', _.noop)
          .sendEvent('start')
          .destroy();
  });
});
describe('.isRunning()', function() {
  it('should check if monitor is running', async () => {
    assert.strictEqual(tester.isRunning(), false);
    tester.start();
    assert.strictEqual(tester.isRunning(), true);
    tester.stop();
    assert.strictEqual(tester.isRunning(), false);
  });
});


os-monitor.d.ts
declare const os: any, events: any, stream: any, _: any, critical: number, defaults: Function;
declare class Monitor extends stream.Readable {
    constructor();
    get version(): string;
    get constants(): MonitorConstants;
    os: any;
    _: any;
    private _monitorState;
    private _read;
    sendEvent(event: string, obj?: InfoObject): Monitor;
    private _cycle;
    start(options?: ConfigObject): Monitor;
    stop(): Monitor;
    reset(): Monitor;
    destroy(err?: any): Monitor;
    config(options?: ConfigObject): ConfigObject;
    isRunning(): boolean;
    private _isEnded;
    throttle(event: string, handler: Function, wait: number): Monitor;
    unthrottle(event: string, handler: Function): Monitor;
    when(event: string): Promise<Thenable> | Thenable;
    private _sanitizeNumber;
    seconds(n: number): number;
    minutes(n: number): number;
    hours(n: number): number;
    days(n: number): number;
}
declare class Thenable extends events.EventEmitter {
    constructor();
    static constants: {
        state: {
            PENDING: string;
            FULFILLED: string;
            REJECTED: string;
        };
    };
    private _thenableState;
    resolve(result: EventObject): Thenable;
    reject(error: any): Thenable;
    then(onFulfilled: Function | undefined, onRejected: Function | undefined): void;
    catch(onRejected: Function | undefined): void;
    private _callOnFulfilled;
    private _callOnRejected;
}
interface ConfigObject {
    delay?: number;
    critical1?: number;
    critical5?: number;
    critical15?: number;
    freemem?: number;
    uptime?: number;
    silent?: boolean;
    stream?: boolean;
    immediate?: boolean;
    throttle?: number;
}
interface MonitorState {
    running: boolean;
    ended: boolean;
    streamBuffering: boolean;
    interval: any;
    config: ConfigObject;
    throttled: Array<{
        originalFn: Function;
        throttledFn: Function;
    }>;
}
interface MonitorConstants {
    events: {
        MONITOR: string;
        UPTIME: string;
        FREEMEM: string;
        LOADAVG1: string;
        LOADAVG5: string;
        LOADAVG15: string;
        START: string;
        STOP: string;
        CONFIG: string;
        RESET: string;
        DESTROY: string;
    };
}
interface InfoObject {
    loadavg?: Array<number>;
    uptime?: number;
    freemem?: number;
    totalmem?: number;
    options?: ConfigObject;
}
interface EventObject extends InfoObject {
    eventType: string;
    timestamp: number;
}


spec-runner.js


const assert = require('assert'),
    os = require('os'),
    mock = require('mock-os'),
    EOL = os.EOL;

// mock os module
mock({
  cpus: [{}, {}]
});

function log(output = '') {
  process.stdout.write(`${output}${EOL}`);
}

const monitor = require('./os-monitor');

log("running tests...");
log();

(() => {

  // API signature
  assert.ok(monitor._read, "internal ._read() method expected");
  assert.ok(monitor.version, "version property expected");
  assert.ok(monitor.sendEvent, ".sendEvent() method expected");
  assert.ok(monitor.start, ".start() method expected");
  assert.ok(monitor.stop, ".stop() method expected");
  assert.ok(monitor.reset, ".reset() method expected");
  assert.ok(monitor.destroy, ".destroy() method expected");
  assert.ok(monitor.config, ".config() method expected");
  assert.ok(monitor.isRunning, ".isRunning() method expected");
  assert.ok(monitor.throttle, ".throttle() method expected");
  assert.ok(monitor.unthrottle, ".unthrottle() method expected");
  assert.ok(monitor.seconds, ".seconds() method expected");
  assert.ok(monitor.minutes, ".minutes() method expected");
  assert.ok(monitor.hours, ".hours() method expected");
  assert.ok(monitor.days, ".days() method expected");
  assert.ok(monitor.Monitor, "Monitor class expected");
  assert.ok(monitor.os, "os object reference expected");
  assert.ok(monitor._, "_ object reference expected");
  assert.ok(monitor.constants, "constants object expected");
  //assert.ok(monitor.whetever, "whatever expected");


  log("API signature OK");


  // config
  monitor.config({test: 123});
  assert.strictEqual(monitor.config().test, 123, "config() : same config expected");
  monitor.start({test: 456});
  assert.strictEqual(monitor.config().test, 456, "start() : same config expected");
  monitor.stop();

  log("config OK");


  // events
  let trace = {};

  monitor.on('monitor', (event) => {
    trace.monitorEvent = event.type;
  });
  monitor.on('loadavg1', (event) => {
    trace.loadavg1Event = event.type;
  });
  monitor.on('loadavg5', (event) => {
    trace.loadavg5Event = event.type;
  });
  monitor.on('loadavg15', (event) => {
    trace.loadavg15Event = event.type;
  });
  monitor.on('freemem', (event) => {
    trace.freememEvent = event.type;
  });
  monitor.on('uptime', (event) => {
    trace.uptimeEvent = event.type;
  });
  monitor.sendEvent('monitor', {type: 'monitor'});
  monitor.sendEvent('loadavg1', {type: 'loadavg1'});
  monitor.sendEvent('loadavg5', {type: 'loadavg5'});
  monitor.sendEvent('loadavg15', {type: 'loadavg15'});
  monitor.sendEvent('freemem', {type: 'freemem'});
  monitor.sendEvent('uptime', {type: 'uptime'});
  
  assert.strictEqual(trace.monitorEvent, "monitor", "'monitor' event expected");
  assert.strictEqual(trace.loadavg1Event, "loadavg1", "'loadavg1' event expected");
  assert.strictEqual(trace.loadavg5Event, "loadavg5", "'loadavg5' event expected");
  assert.strictEqual(trace.loadavg15Event, "loadavg15", "'loadavg15' event expected");
  assert.strictEqual(trace.freememEvent, "freemem", "'freemem' event expected");
  assert.strictEqual(trace.uptimeEvent, "uptime", "'uptime' event expected");
  
  log("events OK");

  monitor.removeAllListeners();
  

  // readable stream interface
  monitor.config({stream: true});
  monitor.sendEvent('monitor', {type: 'monitor'});
  assert.ok(monitor.read(), "output expected from readable stream interface");

  log("readable stream interface OK");


  // isRunning() test
  assert.strictEqual(monitor.isRunning(), false, "isRunning() === false expected");
  monitor.start();
  assert.strictEqual(monitor.isRunning(), true, "isRunning() === true expected");
  monitor.stop();
  assert.strictEqual(monitor.isRunning(), false, "isRunning() === false expected");

  log("isRunning() test OK");
    
    
  // reset() test
  monitor.config({
                 delay: 1000,
                 critical1: 1.5,
                 critical5: 1.5,
                 critical15: 1.5,
                 freemem: 1,
                 uptime: 1,
                 silent: true,
                 stream: true,
                 immediate: true
         });
  monitor.reset();

  assert.strictEqual(monitor.config().delay, 3000, "config should be reset");
  assert.strictEqual(monitor.config().critical1, os.cpus().length, "config should be reset");
  assert.strictEqual(monitor.config().critical5, os.cpus().length, "config should be reset");
  assert.strictEqual(monitor.config().critical15, os.cpus().length, "config should be reset");
  assert.strictEqual(monitor.config().freemem, 0, "config should be reset");
  assert.strictEqual(monitor.config().uptime, 0, "config should be reset");
  assert.strictEqual(monitor.config().silent, false, "config should be reset");
  assert.strictEqual(monitor.config().stream, false, "config should be reset");
  assert.strictEqual(monitor.config().immediate, false, "config should be reset");

  log("reset() test OK");
    
    
  // destroy() test
  monitor.on('close', () => {
    trace.closeEvent = 'close';
  });

  monitor.destroy();

  assert.strictEqual(monitor.isRunning(), false, "isRunning() === false expected");
  assert.strictEqual(trace.closeEvent, "close", "'close' event expected");
  assert.throws(() => {
                    monitor.start();
                },
                "start() should throw");
  assert.throws(() => {
                    monitor.push('some data');
                },
                "push() should throw");

  log("destroy() test OK");

  monitor.removeAllListeners();

})();


// mock tests with 2 fake cpus
mock({
  cpus: [{}, {}],
  loadavg: [3, 3, 3],
  freemem: 1000,
  totalmem: 5000,
  uptime: 10000
});

let monitor2 = new monitor.Monitor();
let monitor3 = new monitor.Monitor();
let trace2 = {};
let trace3 = {};

monitor2.on('freemem', (event) => {
  trace2.freememAbsolute = event;
})
.on('uptime', (event) => {
  trace2.uptime = event;
})
.on('loadavg1', (event) => {
  trace2.loadavg1 = event;
})
.on('loadavg5', (event) => {
  trace2.loadavg5 = event;
})
.on('loadavg15', (event) => {
  trace2.loadavg15 = event;
});

monitor3.on('freemem', (event) => {
  trace3.freememAbsolute = event;
})
.on('uptime', (event) => {
  trace3.uptime = event;
})
.on('loadavg1', (event) => {
  trace3.loadavg1 = event;
})
.on('loadavg5', (event) => {
  trace3.loadavg5 = event;
})
.on('loadavg15', (event) => {
  trace3.loadavg15 = event;
});

monitor2.start({
  freemem: 20000,
  uptime: 1000,
  immediate: true
});

monitor3.start({
  freemem: 200,
  uptime: 1000000,
  critical1: 4,
  critical5: 4,
  critical15: 4,
  immediate: true
});

process.nextTick(() => {
  monitor2.removeAllListeners();
  monitor3.removeAllListeners();
  
  monitor2.on('freemem', (event) => {
    trace2.freememPct = event;
  });
  monitor3.on('freemem', () => {
    trace3.freememPct = event;
  });
});

monitor2.start({
  freemem: 0.7,
  immediate: true
}).stop();

monitor3.start({
  freemem: 0.1,
  immediate: true
}).stop();

setImmediate(() => {
  assert.ok(trace2.freememAbsolute, "freememAbsolute expected");
  assert.ok(trace2.freememPct, "freememPct expected");
  assert.ok(trace2.uptime, "uptime expected");
  assert.ok(trace2.loadavg1, "loadavg1 expected");
  assert.ok(trace2.loadavg5, "loadavg5 expected");
  assert.ok(trace2.loadavg15, "loadavg15 expected");
  
  assert.ok(trace2.freememAbsolute.timestamp, "event.timestamp expected");
  assert.ok(trace2.freememPct.timestamp, "event.timestamp expected");
  assert.ok(trace2.uptime.timestamp, "event.timestamp expected");
  assert.ok(trace2.loadavg1.timestamp, "event.timestamp expected");
  assert.ok(trace2.loadavg5.timestamp, "event.timestamp expected");
  assert.ok(trace2.loadavg15.timestamp, "event.timestamp expected");
  
  assert.ok(trace2.freememAbsolute.loadavg, "event.loadavg expected");
  assert.ok(trace2.freememPct.loadavg, "event.loadavg expected");
  assert.ok(trace2.uptime.loadavg, "event.loadavg expected");
  assert.ok(trace2.loadavg1.loadavg, "event.loadavg expected");
  assert.ok(trace2.loadavg5.loadavg, "event.loadavg expected");
  assert.ok(trace2.loadavg15.loadavg, "event.loadavg expected");
  
  assert.ok(trace2.freememAbsolute.freemem, "event.freemem expected");
  assert.ok(trace2.freememPct.freemem, "event.freemem expected");
  assert.ok(trace2.uptime.freemem, "event.freemem expected");
  assert.ok(trace2.loadavg1.freemem, "event.freemem expected");
  assert.ok(trace2.loadavg5.freemem, "event.freemem expected");
  assert.ok(trace2.loadavg15.freemem, "event.freemem expected");
  
  assert.ok(trace2.freememAbsolute.totalmem, "event.totalmem expected");
  assert.ok(trace2.freememPct.totalmem, "event.totalmem expected");
  assert.ok(trace2.uptime.totalmem, "event.totalmem expected");
  assert.ok(trace2.loadavg1.totalmem, "event.totalmem expected");
  assert.ok(trace2.loadavg5.totalmem, "event.totalmem expected");
  assert.ok(trace2.loadavg15.totalmem, "event.totalmem expected");
  
  assert.ok(trace2.freememAbsolute.uptime, "event.uptime expected");
  assert.ok(trace2.freememPct.uptime, "event.uptime expected");
  assert.ok(trace2.uptime.uptime, "event.uptime expected");
  assert.ok(trace2.loadavg1.uptime, "event.uptime expected");
  assert.ok(trace2.loadavg5.uptime, "event.uptime expected");
  assert.ok(trace2.loadavg15.uptime, "event.uptime expected");
  
  assert.strictEqual(trace2.freememAbsolute.type, "freemem", "event.type === 'freemem' expected");
  assert.strictEqual(trace2.freememPct.type, "freemem", "event.type === 'freemem' expected");
  assert.strictEqual(trace2.uptime.type, "uptime", "event.type === 'uptime' expected");
  assert.strictEqual(trace2.loadavg1.type, "loadavg1", "event.type === 'loadavg1' expected");
  assert.strictEqual(trace2.loadavg5.type, "loadavg5", "event.type === 'loadavg5' expected");
  assert.strictEqual(trace2.loadavg15.type, "loadavg15", "event.type === 'loadavg15' expected");
  
  assert.ok(!trace3.freememAbsolute, "freememAbsolute not expected");
  assert.ok(!trace3.freememPct, "freememPct not expected");
  assert.ok(!trace3.uptime, "uptime not expected");
  assert.ok(!trace3.loadavg1, "loadavg1 not expected");
  assert.ok(!trace3.loadavg5, "loadavg5 not expected");
  assert.ok(!trace3.loadavg15, "loadavg15 not expected");
  
  log("mock tests OK");
});

// throttle
let eventCount = 0;
let monitor4 = new monitor.Monitor();

monitor4.start();
monitor4.throttle('loadavg1', (event) => {
  eventCount++;
}, 100);

monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
setImmediate(() => {
  monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
});
setTimeout(() => {
  monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
  monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
  monitor4.sendEvent('loadavg1', {type: 'loadavg1'});
  setImmediate(() => {
    assert.strictEqual(eventCount, 3, "throttle: eventCount === 3 expected");
    log("throttle() test OK");
    monitor4.stop();
  });
}, 200);

// unthrottle
mock({freemem: 1000});
let eventCount2 = 0;
let monitor5 = new monitor.Monitor();
let handler = function(event) {
  eventCount2++;
};

monitor5.throttle('freemem', handler, 1);
monitor5.throttle('freemem', handler, 1);
monitor5.throttle('freemem', handler, 1);
monitor5.start({'freemem': 10000, delay: 100, immediate: false});

setTimeout(() => {
  monitor5.unthrottle('freemem', handler);
  monitor5.emit('freemem');

  assert.strictEqual(eventCount2, 3, "unthrottle: eventCount === 3 expected");
  log("unthrottle() test OK");
  monitor5.stop();
  mock.restore();
}, 200);
